pkg/neotest/coverage.go

```diff
@@ -0,0 +1,153 @@
+package neotest
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/nspcc-dev/neo-go/pkg/compiler"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+	"github.com/nspcc-dev/neo-go/pkg/vm"
+	"github.com/nspcc-dev/neo-go/pkg/vm/opcode"
+)
+
+var rawCoverage = make(map[scriptHash]*scriptRawCoverage)
+
+var enabled = false
+var coverProfile = ""
+
+type scriptHash = util.Uint160
+
+type scriptRawCoverage struct {
+	debugInfo      *compiler.DebugInfo
+	offsetsVisited []int
+}
+
+type coverBlock struct {
+	startLine uint // Line number for block start.
+	startCol  uint // Column number for block start.
+	endLine   uint // Line number for block end.
+	endCol    uint // Column number for block end.
+	stmts     uint // Number of statements included in this block.
+	counts    uint
+}
+
+type documentName = string
+
+func isCoverageEnabled() bool {
+	if enabled {
+		return true
+	}
+	const coverProfileFlag = "test.coverprofile"
+	flag.VisitAll(func(f *flag.Flag) {
+		if f.Name == coverProfileFlag && f.Value != nil {
+			enabled = true
+			coverProfile = f.Value.String()
+		}
+	})
+	if enabled {
+		// this is needed so go cover tool doesn't overwrite
+		// the file with our coverage when all tests are done
+		flag.Set(coverProfileFlag, "")
+	}
+	return enabled
+}
+
+func coverageHook() vm.OnExecHook {
+	return func(sh scriptHash, offset int, opcode opcode.Opcode) {
+		if cov, ok := rawCoverage[sh]; ok {
```

Comment:
Unprotected access to globals prevents us from executing tests in parallel. What about having a mutex (or `sync.Map`)?

Reply:
I think we just need to use normal Mutex, because there is no atomic "load and update" operation on sync.Map. If multiple goroutines read the map at the same time then `append` will be called from different goroutines too and only 1 value will be added likely.

Reply:
Worth to update the implementation, it's a frequent path.